proposals/0055-transact-without-require-weight-at-most.md

```diff
@@ -0,0 +1,65 @@
+---
+Title: Transact without `require_weight_at_most`
+Number: 55
+Status: Draft
+Version: 0
+Authors:
+ - Adrian Catangiu
+ - Francisco Aguirre
+Created: 2024-04-25
+Impact: High
+Requires:
+Replaces:
+---
+
+## Summary
+
+The Transact instruction shouldn't need the `require_weight_at_most` parameter, since the decoder could just get the weight of the call from its dispatch info.
+
+## Motivation
+
+The UX of using `Transact` is not great, and one part of the problem is guesstimating this `require_weight_at_most`.
+We've seen multiple `Transact`s on-chain failures caused by the "incorrect" use or the parameter. In practice,
+this parameter only adds UX overhead. The vast majority of use cases fall in one of two categories:
+1. Unpaid execution of Transacts - in these cases the `require_weight_at_most` is not really useful, caller doesn't 
+have to pay for it, and on the call site it either fits the block or not;
+2. Paid execution of _single_ Transact - the weight to be spent by the Transact is already covered by the `BuyExecution`
+weight limit parameter.
+
+We've had multiple OpenGov `root/whitelisted_caller` proposals initiated by core-devs completely or partially fail
+because of incorrect configuration of `require_weight_at_most` parameter. This is a strong indication that the
+instruction is hard to use.
+
+## Specification
+
+The proposal is simple: change `Transact` instruction:
+
+```diff
+- Transact { origin_kind: OriginKind, require_weight_at_most: Weight, call: DoubleEncoded<Call> },
++ Transact { origin_kind: OriginKind, maybe_weight_limit: Option<WeightLimit>, call: DoubleEncoded<Call> },
+```
+
+With this new API, users who do not need to artificially limit the maximum weight used by the inner `call`,
+can pass `maybe_weight_limit: None`; while those who need to do it, still can.
+
+## Security considerations
+
+TODO
+
+## Impact
+
+- Impact limited to single instruction: `Transact`.
+- It is an API breaking change for `Transact`. There is a clean path for version upgrading or downgrading of
+`Transact`, but UIs will have to adapt and render the new API.
+
+## Alternatives
+
+Q1: What other designs have been considered and what is the rationale for not choosing them?
+- No alternatives considered (yet).
```

Comment:
Generally I would have assumed that the execution could be replaced by a different executor that ignores the weight limit.

Reply:
yes, but that's an implementation decision, not an `xcm-format` decision

the format currently requires the user to pass in a `require_weight_at_most: Weight` parameter - don't we want to change that too? leaving it there but ignoring looks like technical debt and/or tribal knowledge workaround..

Reply:
Yeah just wanted to ask/ensure. Especially to not requiring to wait for XCM v5.